### ISSUE TITLE
Can't bind simple call-back

### ISSUE THREAD

**Issue by v-yatsenko** [RosettaCommons/binder#62]
It was added a support for std::function.
But when trying to bind a code like following:
void set_callback(std::function(void()) callback)

the binder is crashing on assert (if built as Debug):
build/llvm-4.0.0/tools/clang/include/clang/AST/TemplateBase.h:251: clang::QualType clang::TemplateArgument::getParamTypeForDecl() const: Assertion `getKind() == Declaration && "Unexpected kind"' failed.

Or skips this function if built as Release.

While debugging I have found that it is crashing in class.cpp:is_std_function_bindable()

May suspect that the issue will be also with a call-backs uses built-in types like:
int(); void(int); etc.

**Comment by v-yatsenko** [comment]
Proposing following change:
```
// check if given CXXRecordDecl (which known to be template specialization of std::function) is bindable
bool is_std_function_bindable(CXXRecordDecl const *C)
{
        //outs() << "is_std_function_bindable( " << class_qualified_name(C) << "\n";
        if( auto t = dyn_cast<ClassTemplateSpecializationDecl>(C) ) {
                for(uint i=0; i < t->getTemplateArgs().size(); ++i) {
                        if (t->getTemplateArgs()[i].getKind() == TemplateArgument::Declaration) {
                                //outs() << " template argument: " << template_argument_to_string(t->getTemplateArgs()[i]) << "\n";
                                //t->getTemplateArgs()[i].dump();
                                QualType qt = t->getTemplateArgs()[i].getParamTypeForDecl();
                                //qt.dump();

                                if( FunctionProtoType const *ft = dyn_cast<FunctionProtoType>( qt.getTypePtr() ) ) {
                                        if( not is_bindable( ft->getReturnType() ) ) return false;
                                        for(uint i=0; i < ft->getNumParams(); ++i) {
                                                if( not is_bindable( ft->getParamType(i) ) ) return false;
                                        }
                                }
                        }
                }
        }
        return true;
}
```

**Comment by lyskov** [comment]
Thank you for finding and submitting this @v-yatsenko ! Could you please create a PR with proposed changes so it is clear what exactly your change set is altering? Thanks,

**Comment by lyskov** [comment]
Also, @v-yatsenko could you please post the exact code that trigger this? I have tried to bind:
```
void foo1(std::function< void() > ) {}
void foo2(std::function< int() > ) {}
void foo3(std::function< void(int) > ) {}
```
and could not replicate this issue (using Binder debug clang build). Thanks,

**Comment by v-yatsenko** [comment]
Built a binder like this:
python3 build.py --binder-debug

Please see attached 'Hello World!' project. It is reproducible.
[binder_hw.zip](https://github.com/RosettaCommons/binder/files/2840944/binder_hw.zip)

**Comment by lyskov** [comment]
@v-yatsenko could you please post your system info here: what compiler have you used to build Binder as well as OS info. Also, could you please cut/paste C++ code that trigger this error so it could seen in this issue discussion (GitHub does not display content of zip files etc)? Thanks,

**Comment by lyskov** [comment]
Ok, i looked at your examples (still please post them here) and i still can not reproduce the issue. However your patch look reasonable to me so i will push it anyway. Again: thank you for reporting this!

**Comment by lyskov** [comment]
fixed in a9cfcdc548b5e5134d97a327ccd4bbec07c785a4

**Comment by v-yatsenko** [comment]
Thank you for support.
Info as you've requested.
My system is LinuxMint 19.
GCC v7.3.0

Code that triggers an issue is following:
```
class HW {
public:
  HW() {}
  ~HW() {}

  void subscribe(std::function<void()> f); <<< here a crash occurred when parsing
};
```
